lib/composite_primary_keys/composite_predicates.rb

```diff
@@ -51,9 +51,43 @@ def cpk_join_predicate(table1, key1, table2, key2)
     end
 
     def cpk_in_predicate(table, primary_keys, ids)
-      and_predicates = ids.map do |id|
-        cpk_id_predicate(table, primary_keys, id)
+      if primary_keys.length == 2
+        low_cardinality_key_part, high_cardinality_key_part = if ids.map(&:second).uniq.size > ids.map(&:first).uniq.size
+                                                                %i[first second]
+                                                              else
+                                                                %i[second first]
+                                                              end
+
+        groups = ids.group_by(&low_cardinality_key_part).transform_values do |values|
+          values.map(&high_cardinality_key_part)
+        end
+
+        and_predicates = groups.map do |low_cardinality_value, high_cardinality_values|
+          in_clause = table[primary_keys.send(high_cardinality_key_part)].in(high_cardinality_values.compact)
+          inclusion_clauses = if high_cardinality_values.include?(nil)
+                                Arel::Nodes::Grouping.new(
+                                  Arel::Nodes::Or.new(
+                                    in_clause,
+                                    table[primary_keys.send(high_cardinality_key_part)].eq(nil)
+                                  )
+                                )
+                              else
+                                in_clause
+                              end
+
+          Arel::Nodes::And.new(
+            [
+              table[primary_keys.send(low_cardinality_key_part)].eq(low_cardinality_value),
+              inclusion_clauses
+            ]
+          )
+        end
+      else
+        and_predicates = ids.map do |id|
```

Comment:
I'd probably split this method into 3 methods:

  cpk_in_predicate
  cpk_in_predicate_pairs
  cpk_in_predicate_tuples (not a very good name, have a better idea?)